Add a blog post about decoupled taint eviction controller (#43676)
kubernetes/website: content/en/blog/_posts/2023-12-19-taint-eviction-controller.md

---
layout: blog
title: "Kubernetes 1.29: Decouple taint-manager from node-lifecycle-controller"
date: 2023-12-19
slug: kubernetes-1-29-taint-eviction-controller
---

**Authors:** Yuan Chen (Apple), Andrea Tosatto (Apple)

This blog discusses a new feature in Kubernetes 1.29 to improve the handling of taint-based pod eviction.

## Background

In Kubernetes 1.29, an improvement has been introduced to enhance the handling of taint-based pod eviction on nodes. This blog discusses the changes made to node-lifecycle-controller to separate its responsibilities and improve overall code maintainability.

### Summary of changes

node-lifecycle-controller previously combined two independent functions:

> - Adding a pre-defined set of `NoExecute` taints to nodes based on the node conditions.

Reviewer: If you are proposing a next commit, please change 'node' to 'Node' if you are referring to an API object.

- Performing pod eviction on `NoExecute` taint.

With the Kubernetes 1.29 release, the taint-based eviction implementation has been
moved out of node-lifecycle-controller into a separate and independent component called taint-eviction-controller.
This separation aims to disentangle code, enhance code maintainability,
and facilitate future extensions to either component.

As part of the change, additional metrics are introduced to monitor taint-based pod evictions.

> - `pod_deletion_duration_seconds` measures th latency between the time when a taint effect has been activated for the Pod and its deletion via taint-eviction-controller.

Reviewer: th -> the

- `pod_deletions_total` reports the total number of Pods deleted by taint-eviction-controller since its start.

### How to use the new feature?

> A new feature gate, `SeparateTaintEvictionController`, has been added. The feature is enabled by default as Beta in Kubernetes 1.29.  

Reviewer: We have already documented this gate, right?
Adding a link to the gate page would be desirable.

Author: Added a link to the feature gate reference. Note that a doc. update PR for the change is under review too. https://github.com/kubernetes/website/pull/43555/files.

When this feature is enabled, users can optionally disable taint-based eviction by setting `--controllers=-taint-eviction-controller`
in kube-controller-manager.

> To disable the new feature and use the old  taint-manager within node-lifecylecycle-controller , users can set `SeparateTaintEvictionController=false`. 

Reviewer: This is not a complete sentence considering that not everyone understands that you are talking about a feature gate rather than a command line flag.

Author: Changed it to "users can set the feature gate `SeparateTaintEvictionController=false`".

---
layout: blog
title: "Kubernetes 1.29: Decouple taint-manager from node-lifecycle-controller"
date: 2023-12-19
slug: kubernetes-1-29-taint-eviction-controller
---

**Authors:** Yuan Chen (Apple), Andrea Tosatto (Apple)

This blog discusses a new feature in Kubernetes 1.29 to improve the handling of taint-based pod eviction.

## Background

In Kubernetes 1.29, an improvement has been introduced to enhance the handling of taint-based pod eviction on nodes. This blog discusses the changes made to node-lifecycle-controller to separate its responsibilities and improve overall code maintainability.

> ### Summary of changes

Reviewer 2: ```suggestion
## Summary of changes
```

node-lifecycle-controller previously combined two independent functions:

> - Adding a pre-defined set of `NoExecute` taints to Node based on the node's condition.

Reviewer: In the previous comment, I suggested to change 'node' to 'Node' when you are referring to an API object. That is a generic style for our docs. I was expecting to check the page thoroughly for this situation, not just this single instance.

Reviewer 2: > When you are generally discussing an API object, use [sentence-style capitalization](https://docs.microsoft.com/en-us/style-guide/text-formatting/using-type/use-sentence-style-capitalization).

&mdash; [style guide](https://kubernetes.io/docs/contribute/style/style-guide/#use-upper-camel-case-for-api-objects)

Writing “node” is OK, @tengqm (and @yuanchen8911) - especially in a blog article where the author's own style is allowed to show through.

Reviewer: Writing "node" is okay sometimes, but it is still desirable to conform to a single style guide. When referencing to an API object, we use the object's kind name. For example, "Deployment" and "deployment", "Container" and "container" mean different things.

In line 20, "Adding ... to Node based on the node's condition" is not good.

Author: @tengqm , would you mind suggesting what it should be,  `Node` for both appearances ? Thanks.

Author: > In the previous comment, I suggested to change 'node' to 'Node' when you are referring to an API object. That is a generic style for our docs. I was expecting to check the page thoroughly for this situation, not just this single instance.

"Pod" and "PersistentVolume" are used. We prefer the lower case "pod" in "taint-based pod eviction".

- Performing pod eviction on `NoExecute` taint.

With the Kubernetes 1.29 release, the taint-based eviction implementation has been
moved out of node-lifecycle-controller into a separate and independent component called taint-eviction-controller.
This separation aims to disentangle code, enhance code maintainability,
and facilitate future extensions to either component.

> As part of the change, additional metrics are introduced to monitor taint-based pod evictions.

Reviewer 2: (a nit)
```suggestion
As part of the change, additional metrics were introduced to help you monitor taint-based pod evictions:
```

- `pod_deletion_duration_seconds` measures the latency between the time when a taint effect has been activated for the Pod and its deletion via taint-eviction-controller.
- `pod_deletions_total` reports the total number of Pods deleted by taint-eviction-controller since its start.

> ### How to use the new feature?

Reviewer 2: ```suggestion
## How to use the new feature?
```

> A new feature gate, `SeparateTaintEvictionController`, has been added. The feature is enabled by default as Beta in Kubernetes 1.29. Please refer to the [feature gate document](https://github.com/kubernetes/website/blob/main/content/en/docs/reference/command-line-tools-reference/feature-gates.md).

Reviewer 2: Actually it should be `/docs/reference/command-line-tools-reference/feature-gates/`
(you could also use a `relref` shortcode; we very rarely do)

 

When this feature is enabled, users can optionally disable taint-based eviction by setting `--controllers=-taint-eviction-controller`
in kube-controller-manager.

To disable the new feature and use the old taint-manager within node-lifecylecycle-controller , users can set the feature gate `SeparateTaintEvictionController=false`.

> ### Use Cases

Reviewer 2: ```suggestion
## Use cases
```

> To disable the new feature and use the old  taint-manager within node-lifecylecycle-controller , users can set the feature fate `SeparateTaintEvictionController=false`. 

Reviewer 3: nit: this sentences has some extra spaces in it.

Also `feature fate` should be `feature gate`.

This new feature will allow cluster administrators to extend and enhance the default
taint-eviction-controller and even replace the default taint-eviction-controller with a
custom implementation to meet different needs. An example is to better support
stateful workloads that use PersistentVolume on local disks.

> ### FAQ

Reviewer 2: ```suggestion
## FAQ
```

> stateful workloads that use PersistentVolume on  local disks. 

Reviewer 3: nit: this sentences has some extra spaces in it.

Author: Removed it.

Author: Fixed it.

**Does this feature change the existing behavior of taint-based pod evictions?**

No, the taint-based pod eviction behavior remains unchanged. If the feature gate
`SeparateTaintEvictionController` is turned off, the legacy node-lifecycle-controller with taint-manager will continue to be used.

**Will enabling/using this feature result in an increase in the time taken by any operations covered by existing SLIs/SLOs?**

No.

**Will enabling/using this feature result in an increase in resource usage (CPU, RAM, disk, IO, ...)?**

The increase in resource usage by running a separate `taint-eviction-controller` will be negligible.

> ### Learn more

Reviewer 2: ```suggestion
## Learn more
```

For more details, refer to the [KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-scheduling/3902-decoupled-taint-manager/README.md).

> ### Acknowledgments

Reviewer 2: ```suggestion
## Acknowledgments
```

to separate its responsibilities and improve overall code maintainability.

## Summary of changes

node-lifecycle-controller previously combined two independent functions:

- Adding a pre-defined set of `NoExecute` taints to Node based on Node's condition.
- Performing pod eviction on `NoExecute` taint.

With the Kubernetes 1.29 release, the taint-based eviction implementation has been
moved out of node-lifecycle-controller into a separate and independent component called taint-eviction-controller.
This separation aims to disentangle code, enhance code maintainability,
and facilitate future extensions to either component.

As part of the change, additional metrics were introduced to help you monitor taint-based pod evictions:

- `pod_deletion_duration_seconds` measures the latency between the time when a taint effect
has been activated for the Pod and its deletion via taint-eviction-controller.
- `pod_deletions_total` reports the total number of Pods deleted by taint-eviction-controller since its start.

## How to use the new feature?

A new feature gate, `SeparateTaintEvictionController`, has been added. The feature is enabled by default as Beta in Kubernetes 1.29.

> Please refer to the [feature gate document](/content/en/docs/reference/command-line-tools-reference/feature-gates.md).

Reviewer 4: I think this should reference the relative path on the website and not on github as suggested here https://github.com/kubernetes/website/pull/43676#discussion_r1401314209

Reviewer 2: Good point.

/lgtm cancel

Author: Replaced it with the short URL.

in kube-controller-manager.

To disable the new feature and use the old taint-manager within node-lifecylecycle-controller , users can set the feature gate `SeparateTaintEvictionController=false`.

## Use cases

This new feature will allow cluster administrators to extend and enhance the default
taint-eviction-controller and even replace the default taint-eviction-controller with a
custom implementation to meet different needs. An example is to better support
stateful workloads that use PersistentVolume on local disks.

## FAQ

**Does this feature change the existing behavior of taint-based pod evictions?**

No, the taint-based pod eviction behavior remains unchanged. If the feature gate
`SeparateTaintEvictionController` is turned off, the legacy node-lifecycle-controller with taint-manager will continue to be used.

**Will enabling/using this feature result in an increase in the time taken by any operations covered by existing SLIs/SLOs?**

No.

**Will enabling/using this feature result in an increase in resource usage (CPU, RAM, disk, IO, ...)?**

The increase in resource usage by running a separate `taint-eviction-controller` will be negligible.

## Learn more

> For more details, refer to the [KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-scheduling/3902-decoupled-taint-manager/README.md).

Reviewer 5: ```suggestion
For more details, refer to the [KEP](http://kep.k8s.io/3902).
```
shouldn't the KEP links be like this?

Reviewer 5: shouldnt the KEP links be mentioned with kep.k8s.io @sftim 
I am not sure here as i have seen mix of short link and long one.

Reviewer 2: It's not something I'd hold back an LGTM for. I do like the short links as they are easy to maintain.

Reviewer 5: understood. just to confirm: the shorts links are recommended  but not stated in docs as a must . right?
